File: src/marketdata/providers/tiingo.py
"""Tiingo end-of-day prices.

One request returns bars, dividends and splits together: `divCash` and
`splitFactor` ride along on each daily row rather than coming from separate
corporate-action endpoints. That is why the ingest driver is organised per
provider rather than per table.

Tiingo publishes an adjusted series (`adjOpen`/`adjClose`/...) which we store as
`vendor_adj` for cross-checking only. Total returns get computed downstream from
the raw bars plus distributions -- see schema.sql.

API: https://www.tiingo.com/documentation/end-of-day
"""

import json

import requests

import config
from model import (
    SERIES_RAW,
    SERIES_VENDOR_ADJ,
    Bar,
    Distribution,
    Payload,
    Split,
)

NAME = "tiingo"
EXT = "json"

_URL = "https://api.tiingo.com/tiingo/daily/{symbol}/prices"

# Tiingo returns only the most recent row unless a start date is given.
_EPOCH = "1900-01-01"

_TIMEOUT = 60


def fetch(symbol: str) -> bytes:
    """Download the full available history for `symbol`."""
    response = requests.get(
        _URL.format(symbol=symbol),
        params={"startDate": _EPOCH, "format": "json"},
        # Header rather than a `token=` query parameter: the key would otherwise
        # end up in the archived payload path, logs and any committed fixture.
        headers={"Authorization": f"Token {config.api_key(NAME)}"},
        timeout=_TIMEOUT,
    )
    if not response.ok:
        raise RuntimeError(
            f"tiingo {symbol}: HTTP {response.status_code} {response.reason}\n"
            f"{response.text[:500]}"
        )
    return response.content


def parse(payload: bytes) -> Payload:
    rows = json.loads(payload)
    if not isinstance(rows, list):
        raise ValueError(f"expected a JSON array, got {type(rows).__name__}")

    out = Payload()
    for row in rows:
        # '1993-01-29T00:00:00.000Z' -> '1993-01-29'
        d = row["date"][:10]

        out.bars.append(
            Bar(
                d=d,
                series_kind=SERIES_RAW,
                open=row["open"],
                high=row["high"],
                low=row["low"],
                close=row["close"],
                volume=row["volume"],
            )
        )
        out.bars.append(
            Bar(
                d=d,
                series_kind=SERIES_VENDOR_ADJ,
                open=row["adjOpen"],
                high=row["adjHigh"],
                low=row["adjLow"],
                close=row["adjClose"],
                volume=row["adjVolume"],
            )
        )

        # Tiingo reports a single cash amount with no breakdown into ordinary
        # income vs. capital gains, so everything lands in 'unknown'. The issuer
        # files are what carry the categories; reconciliation compares this lump
        # sum against their sum for the same ex-date.
        if (amount := row["divCash"]) != 0.0:
            out.distributions.append(
                Distribution(ex_date=d, kind="unknown", amount=amount)
            )

        # 2.0 for a 2-for-1. Stored as num/den to keep the ratio exact.
        if (factor := row["splitFactor"]) != 1.0:
            out.splits.append(Split(ex_date=d, num=factor, den=1.0))

    return out

File: src/atlib/atlib/marketdata/tiingo.cpp
#include <atlib/marketdata/tiingo.h>

#include <atlib/marketdata/credentials.h>
#include <atlib/net/download.h>

#include <rfl/json.hpp>

#include <cctype>
#include <charconv>

namespace tiingo
{
namespace
{

constexpr string_view k_prices_url = "https://api.tiingo.com/tiingo/daily/{}/prices";

// Tiingo returns only the most recent row unless a start date is given. This one
// predates every listing, so "the full history" needs no per-symbol knowledge.
constexpr string_view k_epoch = "1900-01-01";

constexpr chr::seconds k_timeout{60};

// Enough of a failing response to identify the failure, not so much that a stray
// HTML error page fills the log.
constexpr size_t k_error_body_limit = 500;

// The symbol is interpolated into the URL *path*, where a stray character does
// not merely fail: "../.." climbs out of /daily/ and addresses some other
// endpoint entirely, and a '?' or '#' truncates the path. Real tickers are
// alphanumeric with dots and dashes, so anything else is rejected rather than
// escaped -- no legitimate symbol needs escaping, and refusing is easier to
// reason about than encoding.
bool is_usable_symbol(string_view symbol)
{
    return !symbol.empty() && ra::all_of(symbol, [](const char c) {
        return std::isalnum(static_cast<unsigned char>(c)) != 0 || c == '.' || c == '-';
    });
}

// One row of the response, as it arrives.
//
// The adjOpen/adjHigh/adjLow/adjClose/adjVolume fields are deliberately absent:
// reflect-cpp ignores what nobody declares, and equity.h explains why the
// adjusted series is not wanted. Member names are snake_case and the reader is
// given SnakeCaseToCamelCase, so div_cash reads divCash off the wire without
// the struct having to adopt the vendor's spelling.
struct Row {
    string date;
    double open;
    double high;
    double low;
    double close;
    double volume;
    double div_cash;
    double split_factor;
};

// "1993-01-29T00:00:00.000Z" -> 1993-01-29.
//
// Only the date is meaningful. The time is always midnight UTC, which is not when
// anything traded, and is discarded rather than interpreted.
expected<chr::local_days, string> parse_date(string_view text)
{
    // Reject rather than repair. A date is the one field where a wrong-but-
    // plausible value cannot be spotted downstream: it would just quietly place
    // a bar on the wrong day.
    const auto malformed = [text] {
        return unexpected(format("not a date: \"{}\"", text));
    };

    if (text.size() < 10 || text[4] != '-' || text[7] != '-') {
        return malformed();
    }

    const auto number = [text](size_t offset, size_t length) -> optional<int> {
        int value = 0;
        const char* begin = text.data() + offset;
        const auto [end, ec] = std::from_chars(begin, begin + length, value);
        if (ec != std::errc{} || end != begin + length) {
            return nullopt;
        }
        return value;
    };

    const auto year = number(0, 4);
    const auto month = number(5, 2);
    const auto day = number(8, 2);
    if (!year || !month || !day) {
        return malformed();
    }

    const chr::year_month_day date{
      chr::year{*year}, chr::month{static_cast<unsigned>(*month)}, chr::day{static_cast<unsigned>(*day)}
    };
    if (!date.ok()) {
        return malformed();
    }

    return chr::local_days{date};
}

} // namespace

expected<string, string> fetch(string_view symbol)
{
    if (!is_usable_symbol(symbol)) {
        return unexpected(format("not a usable symbol: \"{}\"", symbol));
    }

    const auto key = api_key(k_name);
    if (!key) {
        return unexpected(key.error());
    }

    const auto response = http_get({
      .url = format(k_prices_url, symbol),
      .query = {{"startDate", string(k_epoch)}, {"format", "json"}},
      // A header rather than a token= query parameter: a token in the query
      // would end up in the cached payload's path, in logs, and in every error
      // message that quotes the request.
      .headers = {{"Authorization", format("Token {}", *key)}},
      .timeout = k_timeout,
    });
    if (!response) {
        return unexpected(response.error());
    }

    if (!response->ok()) {
        // The body carries the reason -- a bad key, an unknown ticker, a
        // quota -- and the status alone never distinguishes them.
        return unexpected(format(
          "tiingo {}: HTTP {}\n{}", symbol, response->status, string_view(response->body).substr(0, k_error_body_limit)
        ));
    }

    return response->body;
}

expected<EquityHistory, string> parse(string_view symbol, string_view payload)
{
    const auto rows = rfl::json::read<vector<Row>, rfl::SnakeCaseToCamelCase>(payload);
    if (!rows) {
        return unexpected(format("tiingo {}: {}", symbol, rows.error().what()));
    }

    EquityHistory history;
    history.symbol = string(symbol);
    history.bars.reserve(rows->size());

    for (const auto& row : *rows) {
        const auto date = parse_date(row.date);
        if (!date) {
            return unexpected(format("tiingo {}: {}", symbol, date.error()));
        }

        history.bars.push_back(
          {.date = *date, .open = row.open, .high = row.high, .low = row.low, .close = row.close, .volume = row.volume}
        );

        // Tiingo reports these on every row, zero and one on the overwhelming
        // majority of them. Keeping the non-events would bury the events.
        if (row.div_cash != 0.0) {
            history.distributions.push_back({.ex_date = *date, .amount = row.div_cash});
        }
        if (row.split_factor != 1.0) {
            history.splits.push_back({.ex_date = *date, .factor = row.split_factor});
        }
    }

    // Tiingo returns rows oldest-first, and everything downstream assumes it.
    // Checking rather than sorting: an unsorted or duplicated response means the
    // endpoint is not behaving as understood, and quietly repairing the order
    // would hide whatever else changed with it.
    const auto out_of_order = ra::adjacent_find(history.bars, [](const DailyBar& a, const DailyBar& b) {
        return a.date >= b.date;
    });
    if (out_of_order != history.bars.end()) {
        return unexpected(format("tiingo {}: rows are not in ascending date order at {}", symbol, out_of_order->date));
    }

    return history;
}

} // namespace tiingo
